---
title: "Qiime2 Basics"
---

QIIME2, with Qiime pronounced as "chime", or Quantitative Insights Into Microbial Ecology 2, is
an open source software package for comparison and analysis of microbial communities.
Culture independent analysis of microbial communities has been enabled by high-throughput
genetic sequencing and high-throughput data analysis. Multiplexing and high-throughput
sequencing technologies, such as Illumina sequencing, allow scientists to perform parallel
sequencing of hundreds of samples at high depths. Open-source bioinformatics pipelines, such
as QIIME2, allow for robust analysis of millions of sequences. While these tools are powerful
and flexible, they are complex and can be difficult to learn.

The core goal of this workshop is to create a standard pipeline that is accessible to first time
QIIME2 users. QIIME2 is extremely flexible and can accommodate various sequencing
technologies and methods of data analysis. This tutorial presents a set of scripts that will allow a
user to quickly progress through “typical” analysis of 16S rRNA gene data. These scripts have
been built for use with Illumina sequencing technology and the Juniata College HHMI
computational environment.

:::{.callout-note}
Take a moment to think about the biological questions you are investigating. Are you interested
in how members of the microbial community are varying with environmental parameters? With
a specific treatment or disease state? Are you interested in the presence / absence, relative
abundance, or both? Are you interested in rare members of the microbial community? Do you
have any specific hypotheses? Take a moment to write these biological questions down in the
space below. This will help you understand what statistical approaches you may want to use.
:::

All of the files that you will use with Qiime2 are either .qza (artifacts) or .qzv (visualizations).
Qiime2 tracks how those files were created, which helps with reproducibility. The website
view.qiime2.org is used to see this information and visualizations.

Artifacts have different [semantic types](https://docs.qiime2.org/2022.11/semantic-types/). These
semantic types are basically analogous to different file types on your computer and help prevent
users from applying analyses incorrectly.

Qiime2’s functionality is based on plugins. All the plugins that you need for the main pipeline are
installed by default. However, you might find some plugins listed [here](https://library.qiime2.org/plugins/) interesting.

The [Qiime2 website](https://docs.qiime2.org/2022.11/) has a variety of tutorials if you want any additional practice or just want to
see what else it can do. The [Moving Pictures tutorial](https://docs.qiime2.org/2022.11/) in particular has a lot of similarities with our pipeline

## Installing Qiime2

We will be installing Qiime2 my creating a miniconda environment, then installing
Qiime2 into that environment. You can do this by running the following commands:

``` bash
wget https://data.qiime2.org/distro/core/qiime2-2022.11-py38-linux-conda.yml
conda env create -n qiime2-2022.11 --file qiime2-2022.11-py38-linux-conda.yml
```

:::{.callout-note}
This may take a while to run.
:::
:::{.callout-note}
If you get a HTTP error, try running the command again.
:::

## Other Necessary Installations

We also need to create another environment; we’ll use this environment with script 6.1 and 7.2:

``` bash
conda env create -n biom-convert --file /home/see/YML_Files/biom-convert5.yml
```

And, one more for script 6.0…

``` bash
conda env create -n decontam --file /home/see/YML_Files/decontam2.yaml
```

These environments can be activated using conda activate [NAME OF ENVIRONMENT]. So, for
example

``` bash
conda activate qiime2-2022.11
```

And, they can be deactivated using the command:

``` bash
conda deactivate
```

You can check that your environments installed correctly by activating them and running a
command that should now work with the --help flag and seeing if that brings up the command’s
options.

Check that the biom-convert environment installed correctly by doing:

``` bash
conda activate biom-convert
```

Alternatively to check, use: 

``` bash
conda env list
```

:::{.callout-note}
Confirm that your environments work. Use the command `biom convert --help` for the
biom-convert environment and the command `qiime tools import --help` for the
qiime2-2020.11 environment.
:::

## Qiime2 Files

Qiime2 uses .qza and .qzv files.
Qiime2 .qza files are referred to as artifacts. These files contain data used in analyses, in
addition to information about how that data were generated (provenance) and the file type,
e.g. FeatureTable[Frequency]. Artifacts can be viewed at [view.qiime2.org](view.qiime2.org).

You can see all of the formats and types by using flags with the qiime tools import command
after activating the environment

``` bash
qiime tools import --show-importable-types
qiime tools import --show-importable-formats
```

![](images/tutorial-images-2/2-1_0.png)

The details tab (shown above) gives the file name, the Qiime2 id (uuid), the type, and format. It
also lists citations for the tools used to create the file below.
The provenance tab (the tab in the blue rectangle above) contains information about how the
data were generated.


![](images/tutorial-images-2/2-1_1.png)

If you click on the circle, that will just show you the uuid, type, and format again. But, if you click
on the square surrounding it…

![](images/tutorial-images-2/2-1_2.png)

You’ll get basically all the information you could ever want about how it was made. As you can
see above, it tells you when the file was made, how long its creation took, what plugin was used
to make it, the inputs used, and even the parameters used for the command. This is great for
keeping track of how Qiime2 data and figures were generated.

Like artifacts, Qiime2 visualizations (.qzv files) also have Details and Provenance tabs when
viewed with [view.qiime2.org.](view.qiime2.org.).

![](images/tutorial-images-2/2-1_3.png)

But, it has another tab, the Visualization tab. And, above, is the same data, except one
additional command was used to convert the table from an artifact to a visualization, making it
more readable and providing you with additional information about it.

![](images/tutorial-images-2/2-1_4.png)

You can see that there’s another square now in the provenance tab. That’s because of the
additional command converting the artifact to a visualization, and on the right side, you can see
that the details about that command are provided.

But, let’s look at the two sets of tabs. The tabs in the purple rectangle are the same for every
visualization and artifact. In other words, all visualizations have Visualization, Details, and
Provenance tabs, while all artifacts have Details and Provenance tabs.

The tabs in the orange box vary based on which tab in the purple area you’re in, and there may
not even be any tabs in this area, as is the case when on the Details tab for the table.qza artifact
above. For the Visualization tab, the tabs in the orange rectangle area also vary depending on
the data being visualized and type of visualization.
